### PR TITLE
nimble/store: Fix nimble store behavior when CCCDs exceed static defined limit

### PR DESCRIPTION
**Pull request by prasad-alatkar**
**Issue reference**: #788 

**Description:**
 In cases where CCCDs are exceeding MYNEWT_VAL(BLE_STORE_MAX_CCCDS) (overflow event), we can end up in continuous while loop.

**Change list:**
1. In `ble_gap_unpair_oldest_peer` changed return value from `0` to `BLE_HS_ENOENT` when there is no entry found.
2. Modified `ble_store_util_status_rr` to address overflow event of CCCDs.

Previous to this PR, if CCCDs used to exceed the MAX_CCCDs then in `ble_store_write` call there used to come OVERFLOW event, which used to call `ble_store_delete_oldest_peer`, now here as we do not have any peer to be deleted, we should have identified that we need to make space for CCCDs and not `peer_sec` or `our_sec`. The PR tries to address this issue.

### CODE REVIEW

**Comment by prasad-alatkar** [comment]
@h2zero That is a valid point, will update if I find anything. Maybe @rymanluk can explain it to us.

File: nimble/host/src/ble_gap.c [hunk 1]
```diff
@@ -5594,7 +5594,7 @@ ble_gap_unpair_oldest_peer(void)
     }
 
     if (num_peers == 0) {
-        return 0;
+        return BLE_HS_ENOENT;
```
Comment (rymanluk) [inline nimble/host/src/ble_gap.c:5597]: Can we make it as separate patch?

File: nimble/host/src/ble_store_util.c [hunk 2]
```diff
@@ -230,13 +299,35 @@ ble_store_util_delete_oldest_peer(void)
...
+    ble_addr_t peer_id_addr;
+    int num_peers;
+
     switch (event->event_code) {
     case BLE_STORE_EVENT_OVERFLOW:
         switch (event->overflow.obj_type) {
             case BLE_STORE_OBJ_TYPE_OUR_SEC:
             case BLE_STORE_OBJ_TYPE_PEER_SEC:
             case BLE_STORE_OBJ_TYPE_CCCD:
-                return ble_gap_unpair_oldest_peer();
```
Comment (rymanluk) [inline nimble/host/src/ble_store_util.c:239]: I believe that for CCCD we should have different handling and if we want to unpair older peer we should not unpair peer for which we store CCCD.  If there is nothing to unpair just return the error.

In the CCCD event there is and ble_addr we should use to make sure that one is no unpaired.

File: nimble/host/src/ble_gap.c (hunk 1 above)
Comment (prasad-alatkar) [reply nimble/host/src/ble_gap.c:5597]: Done !!

File: nimble/host/src/ble_store_util.c (hunk 2 above)
Comment (prasad-alatkar) [reply nimble/host/src/ble_store_util.c:239]: @rymanluk agreed !! I have made relevant changes. For CCCD,  if `ble_gap_unpair_oldest_peer` is unable to delete any peer, we continue to find other peers occupying CCCDs in storage. I have also modified callback function (`ble_store_util_iter_peer_cccd`) to skip the current (i.e. for which CCCDs need to be stored) peer entry.

File: nimble/host/src/ble_store_util.c [hunk 3]
```diff
@@ -230,16 +315,43 @@ ble_store_util_delete_oldest_peer(void)
...
-            case BLE_STORE_OBJ_TYPE_CCCD:
-                return ble_gap_unpair_oldest_peer();
-
-            default:
-                return BLE_HS_EUNKNOWN;
+        case BLE_STORE_OBJ_TYPE_OUR_SEC:
+        case BLE_STORE_OBJ_TYPE_PEER_SEC:
+            return ble_gap_unpair_oldest_peer();
+        case BLE_STORE_OBJ_TYPE_CCCD:
+            if ((rc = ble_gap_unpair_oldest_peer()) == BLE_HS_ENOENT) {
```
Comment (rymanluk) [inline nimble/host/src/ble_store_util.c:329]: I was thinking to call here other version of `unpair_oldest` which would take as a parameter `ble_addr_t ` of device which we don't want to unpair.
If the only device which could be unpair is the one provided, then unpair is not done and this function returns error.

In such a case we know that we cannot store CCC and there is nothing we can do more here. All the other code is not needed then.

Comment (prasad-alatkar) [reply nimble/host/src/ble_store_util.c:329]: @rymanluk That is what I have done actually in `ble_store_util_subscribed_cccds` !! What I think is `ble_gap_unpair_oldest_peer` is not anyway going to unpair current peer (the one for which we want to make space), so introducing similar function with `ble_addr_t` as input won't be really that helpful. Please let me know your thoughts on this.

The issue I believe is only confined to CCCDs, so  `ble_gap_unpair_oldest_peer()) == BLE_HS_ENOENT` will make sure that we do not have any bonded devices and still we are out of storage space, which can be used to conclude that we need to clean old CCCDs (except the current one).

Comment (rymanluk) [reply nimble/host/src/ble_store_util.c:329]: @prasad-alatkar are you sure that `ble_gap_unpair_oldest_peer()` will not unpair current peer? For me it looks it will.

Comment (h2zero) [reply nimble/host/src/ble_store_util.c:329]: I believe @rymanluk is correct, if the current peer exceeds the cccd limit it will likely unpair that peer and disconnect. I like the do nothing if full approach here.

Comment (prasad-alatkar) [reply nimble/host/src/ble_store_util.c:329]: @rymanluk @h2zero , disregard my earlier comment, yes current peer will be unpaired, and we should simply return error in that case.

I was thinking of adding utility function (`ble_store_clean_old_cccd`) which will delete CCCDs of all unbonded peers. This can be called first in case of `CCCD overflow` and if it fails to delete peer then we can continue with  modified `ble_gap_unpair_oldest_peer()`. Please let me know your views on it.

Comment (rymanluk) [reply nimble/host/src/ble_store_util.c]: I believe that `ble_store_util_delete_peer()` deletes all what we need.

Comment (prasad-alatkar) [reply nimble/host/src/ble_store_util.c]: Hi @rymanluk I think for some peers `CCCD` will exist but `peer_sec` and `our_sec` won't exist (Ref: [subscribe-event ble_gatts_clt_cfg_access](https://github.com/apache/mynewt-nimble/blob/9e26fbe97b3d0bb3a22f15ed9c6035be648d1104/nimble/host/src/ble_gatts.c#L782)). So we need mechanism to clean these stored CCCDs, that is why I have introduced `ble_store_clean_old_cccd()` function.

File: nimble/host/src/ble_gap.c [hunk 4]
```diff
@@ -5605,6 +5605,41 @@ ble_gap_unpair_oldest_peer(void)
     return 0;
 }
 
+int
+ble_gap_unpair_oldest_except_curr(const ble_addr_t *curr_peer)
+{
+    ble_addr_t oldest_peer_id_addr[MYNEWT_VAL(BLE_STORE_MAX_BONDS)];
```
Comment (rymanluk) [inline nimble/host/src/ble_gap.c:5611]: nitpick: please rename to `peer_id_addrs`

File: nimble/host/include/host/ble_gap.h [hunk 5]
```diff
@@ -1896,6 +1896,20 @@ int ble_gap_unpair(const ble_addr_t *peer_addr);
...
+ *
+ * @param peer_addr             Address of the current peer (not to be deleted)
+ *
+ * @return                      0 on success;
+ *                              A BLE host HCI return code if the controller
+ *                                  rejected the request;
+ *                              A BLE host core return code on unexpected
+ *                                  error.
+ */
+int ble_gap_unpair_oldest_except_curr(const ble_addr_t *curr_peer);
```
Comment (rymanluk) [inline nimble/host/include/host/ble_gap.h:1911]: let us call it `ble_gap_unpair_oldest_except` () and change the parameter name to `peer_addr`. 
This functionality is not really dedicated only to current device. You might want to have device which you don't want to unpair no matter what :)

File: nimble/host/src/ble_store_util.c [hunk 6]
```diff
@@ -230,16 +372,24 @@ ble_store_util_delete_oldest_peer(void)
...
-                return ble_gap_unpair_oldest_peer();
-
-            default:
-                return BLE_HS_EUNKNOWN;
+        case BLE_STORE_OBJ_TYPE_OUR_SEC:
+        case BLE_STORE_OBJ_TYPE_PEER_SEC:
+            return ble_gap_unpair_oldest_peer();
+        case BLE_STORE_OBJ_TYPE_CCCD:
+            /* Try to remove unbonded CCCDs first */
+            if ((rc = ble_store_clean_old_cccds((void *) &event->overflow.value->cccd.peer_addr)) == BLE_HS_ENOENT) {
```
Comment (rymanluk) [inline nimble/host/src/ble_store_util.c:384]: I don't think we need to clean unbodend CCCDs as we do not keep such in the storage. Unless I miss something, but CCCDs are stored only for bonded devices, and once unbonded, we do clean CCCDs. If it does not work like this (I believe it does) then we have problem somewhere else.

In this case we just need to call the new function and that is it. Nothing else is needed.

File: nimble/host/src/ble_gap.c [hunk 7]
```diff
@@ -5605,6 +5605,41 @@ ble_gap_unpair_oldest_peer(void)
...
+    if (rc != 0) {
+        return rc;
+    }
+
+    if (num_peers == 0) {
+        return BLE_HS_ENOENT;
+    }
+
+    for (i = 0; i < num_peers; i++) {
+        if (memcmp(curr_peer, &oldest_peer_id_addr[i], sizeof (ble_addr_t)) != 0) {
```
[the remark this answers is not in the document]
Comment (prasad-alatkar) [reply nimble/host/src/ble_gap.c:5626]: Sure !!

File: nimble/host/include/host/ble_gap.h (hunk 5 above)
Comment (prasad-alatkar) [reply nimble/host/include/host/ble_gap.h:1911]: Agreed !!

File: nimble/host/src/ble_store_util.c (hunk 6 above)
Comment (prasad-alatkar) [reply nimble/host/src/ble_store_util.c:384]: Hi @rymanluk Please let me know if I am missing something. As per my understanding CCCDs for unbonded devices can get stored as `ble_gatts_clt_cfg_access --> ble_store_write_cccd --> ble_store_write(BLE_STORE_OBJ_TYPE_CCCD, store_value)`.

Comment (h2zero) [reply nimble/host/src/ble_store_util.c:384]: > I don't think we need to clean unbodend CCCDs as we do not keep such in the storage. Unless I miss something, but CCCDs are stored only for bonded devices, and once unbonded, we do clean CCCDs. If it does not work like this (I believe it does) then we have problem somewhere else.
> 
> In this case we just need to call the new function and that is it. Nothing else is needed.

@rymanluk CCCD's do get stored if the devices pair but one or both of them does not accept bonding, so cleaning those up makes sense in this case. Although it would be better if that never happened to begin with.

Comment (rymanluk) [reply nimble/host/src/ble_store_util.c:384]: @h2zero actually CCCD's do get stored when devices are bonded.  That happens after pairing is completed and both devices have bonding flag in set. 

If you do see an issue that CCCDs are stored but device is not actually bonded, please report it as an issue.
